Insert dml encounters `Duplicate entry ` ERROR.

```
[2025/05/13 05:28:39.132 +00:00] [ERROR] [mysql_writer_dml.go:618] ["execute DMLs with error, can not retry"] [error="Error 1062 (23000): Duplicate entry '474448686a6a7a536c42624a71564e79' for key 'contents_0.PRIMARY'"] [duration=3.353µs] [query="UPDATE `db0509`.`contents_0` SET `id`=CASE WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? THEN ? WHEN `id` = ? T"] [count=257] [changefeed=default/test-split]
[2025/05/13 05:28:39.813 +00:00] [ERROR] [event_dispatcher_manager.go:548] ["Event Dispatcher Manager Meets Error"] [changefeedID=default/test-split] [error="Error 1062 (23000): Duplicate entry '47444d584443596859666e784e776f4f' for key 'contents_0.PRIMARY'"] [errorVerbose="Error 1062 (23000): Duplicate entry '47444d584443596859666e784e776f4f' for key 'contents_0.PRIMARY'\[ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/errors.go:178\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/juju_adaptor.go:15\ngithub.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDMLWithMaxRetries.func2\n\tgithub.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_dml.go:523\ngithub.com/pingcap/ticdc/pkg/retry.run\n\tgithub.com/pingcap/ticdc/pkg/retry/retry_with_opt.go:57\ngithub.com/pingcap/ticdc/pkg/retry.Do\n\tgithub.com/pingcap/ticdc/pkg/retry/retry_with_opt.go:34\ngithub.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDMLWithMaxRetries\n\tgithub.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_dml.go:501\ngithub.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).Flush\n\tgithub.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer.go:183\ngithub.com/pingcap/ticdc/downstreamadapter/sink/mysql.(*sink).runDMLWriter\n\tgithub.com/pingcap/ticdc/downstreamadapter/sink/mysql/sink.go:187\ngithub.com/pingcap/ticdc/downstreamadapter/sink/mysql.(*sink).Run.func2\n\tgithub.com/pingcap/ticdc/downstreamadapter/sink/mysql/sink.go:124\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.13.0/errgroup/errgroup.go:79\nruntime.goexit\n\truntime/asm_amd64.s:1700](http://ngithub.com/pingcap/errors.AddStack/n/tgithub.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/errors.go:178/ngithub.com/pingcap/errors.Trace/n/tgithub.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/juju_adaptor.go:15/ngithub.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDMLWithMaxRetries.func2/n/tgithub.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_dml.go:523/ngithub.com/pingcap/ticdc/pkg/retry.run/n/tgithub.com/pingcap/ticdc/pkg/retry/retry_with_opt.go:57/ngithub.com/pingcap/ticdc/pkg/retry.Do/n/tgithub.com/pingcap/ticdc/pkg/retry/retry_with_opt.go:34/ngithub.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDMLWithMaxRetries/n/tgithub.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_dml.go:501/ngithub.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).Flush/n/tgithub.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer.go:183/ngithub.com/pingcap/ticdc/downstreamadapter/sink/mysql.(*sink).runDMLWriter/n/tgithub.com/pingcap/ticdc/downstreamadapter/sink/mysql/sink.go:187/ngithub.com/pingcap/ticdc/downstreamadapter/sink/mysql.(*sink).Run.func2/n/tgithub.com/pingcap/ticdc/downstreamadapter/sink/mysql/sink.go:124/ngolang.org/x/sync/errgroup.(*Group).Go.func1/n/tgolang.org/x/sync@v0.13.0/errgroup/errgroup.go:79/nruntime.goexit/n/truntime/asm_amd64.s:1700)"]
```

--- a/pkg/sink/mysql/mysql_writer_dml.go
+++ b/pkg/sink/mysql/mysql_writer_dml.go
@@ -80,9 +80,7 @@ func (w *Writer) prepareDMLs(events []*commonEvent.DMLEvent) *preparedDMLs {
 			queryList, argsList = w.generateNormalSQLs(eventsInGroup)
 		} else if len(eventsInGroup) > 0 {
 			// if the events are in different safe mode, we can't use the batch dml generate
-			firstEventSafeMode := !w.cfg.SafeMode && eventsInGroup[0].CommitTs > eventsInGroup[0].ReplicatingTs
-			finalEventSafeMode := !w.cfg.SafeMode && eventsInGroup[len(eventsInGroup)-1].CommitTs > eventsInGroup[len(eventsInGroup)-1].ReplicatingTs
-			if firstEventSafeMode != finalEventSafeMode {
+			if shouldSafeMode(w.cfg.SafeMode, eventsInGroup) {
 				queryList, argsList = w.generateNormalSQLs(eventsInGroup)
 			} else {
 				// use the batch dml generate
@@ -99,6 +97,18 @@ func (w *Writer) prepareDMLs(events []*commonEvent.DMLEvent) *preparedDMLs {
 	}
 
 	return dmls
+}
+
+func shouldSafeMode(safemode bool, events []*commonEvent.DMLEvent) bool {
+	if safemode {
+		return true
+	}
+	for _, event := range events {
+		if event.CommitTs < event.ReplicatingTs {
+			return true
+		}
+	}
+	return false
 }
 
 // for generate batch sql for multi events, we first need to compare the rows with the same pk, to generate the final rows.

--- a/pkg/sink/mysql/mysql_writer_dml.go
+++ b/pkg/sink/mysql/mysql_writer_dml.go
@@ -72,20 +72,10 @@ func (w *Writer) prepareDMLs(events []*commonEvent.DMLEvent) *preparedDMLs {
 	)
 	for _, eventsInGroup := range eventsGroup {
 		tableInfo := eventsInGroup[0].TableInfo
-		if !tableInfo.HasPrimaryKey() || tableInfo.HasVirtualColumns() {
-			// check if the table has a handle key or has a virtual column
+		if !shouldGenBatchSQL(tableInfo.HasPrimaryKey(), tableInfo.HasVirtualColumns(), eventsInGroup, w.cfg.SafeMode) {
 			queryList, argsList = w.generateNormalSQLs(eventsInGroup)
-		} else if len(eventsInGroup) == 1 && eventsInGroup[0].Len() == 1 {
-			// if there only one row in the group, we can use the normal sql generate
-			queryList, argsList = w.generateNormalSQLs(eventsInGroup)
-		} else if len(eventsInGroup) > 0 {
-			// if the events are in different safe mode, we can't use the batch dml generate
-			if shouldSafeMode(w.cfg.SafeMode, eventsInGroup) {
-				queryList, argsList = w.generateNormalSQLs(eventsInGroup)
-			} else {
-				// use the batch dml generate
-				queryList, argsList = w.generateBatchSQL(eventsInGroup)
-			}
+		} else {
+			queryList, argsList = w.generateBatchSQL(eventsInGroup)
 		}
 		dmls.sqls = append(dmls.sqls, queryList...)
 		dmls.values = append(dmls.values, argsList...)
@@ -99,6 +89,32 @@ func (w *Writer) prepareDMLs(events []*commonEvent.DMLEvent) *preparedDMLs {
 	return dmls
 }
 
+// shouldGenBatchSQL determines whether batch SQL generation should be used based on table properties and events.
+// Batch SQL generation is used when:
+// 1. The table has a primary key
+// 2. The table doesn't have virtual columns
+// 3. There's more than one row in the group
+// 4. All events have the same safe mode status
+func shouldGenBatchSQL(hasPK bool, hasVirtualCols bool, events []*commonEvent.DMLEvent, safemode bool) bool {
+	if !hasPK || hasVirtualCols {
+		return false
+	}
+
+	if len(events) == 1 && events[0].Len() == 1 {
+		return false
+	}
+
+	if shouldSafeMode(safemode, events) {
+		return false
+	}
+
+	return true
+}
+
+// shouldSafeMode determines if the events should be processed in safe mode.
+// Safe mode is enabled when:
+// 1. The global safe mode is enabled, or
+// 2. Any event's commit timestamp is less than its replicating timestamp
 func shouldSafeMode(safemode bool, events []*commonEvent.DMLEvent) bool {
 	if safemode {
 		return true

--- a/pkg/sink/mysql/mysql_writer_dml.go
+++ b/pkg/sink/mysql/mysql_writer_dml.go
@@ -104,27 +104,41 @@ func shouldGenBatchSQL(hasPK bool, hasVirtualCols bool, events []*commonEvent.DM
 		return false
 	}
 
-	if shouldSafeMode(safemode, events) {
-		return false
-	}
-
-	return true
+	return allRowInSameSafeMode(safemode, events)
 }
 
-// shouldSafeMode determines if the events should be processed in safe mode.
-// Safe mode is enabled when:
-// 1. The global safe mode is enabled, or
-// 2. Any event's commit timestamp is less than its replicating timestamp
-func shouldSafeMode(safemode bool, events []*commonEvent.DMLEvent) bool {
+// allRowInSameSafeMode determines whether all DMLEvents in a batch have the same safe mode status.
+// Safe mode is either globally enabled via the safemode parameter, or determined per event
+// by comparing CommitTs and ReplicatingTs.
+//
+// Parameters:
+//   - safemode: If true, global safe mode is enabled and the function returns true immediately
+//   - events: A slice of DMLEvents to check for consistent safe mode status
+//
+// Returns:
+//
+//	true if either:
+//	- global safe mode is enabled (safemode=true), or
+//	- all events have the same safe mode status (all events' CommitTs > ReplicatingTs, or all ≤)
+//	false if events have inconsistent safe mode status
+func allRowInSameSafeMode(safemode bool, events []*commonEvent.DMLEvent) bool {
 	if safemode {
 		return true
 	}
+
+	if len(events) == 0 {
+		return false
+	}
+
+	firstSafeMode := events[0].CommitTs > events[0].ReplicatingTs
 	for _, event := range events {
-		if event.CommitTs < event.ReplicatingTs {
-			return true
+		currentSafeMode := event.CommitTs > event.ReplicatingTs
+		if currentSafeMode != firstSafeMode {
+			return false
 		}
 	}
-	return false
+
+	return true
 }
 
 // for generate batch sql for multi events, we first need to compare the rows with the same pk, to generate the final rows.

--- a/pkg/sink/mysql/mysql_writer_dml_test.go
+++ b/pkg/sink/mysql/mysql_writer_dml_test.go
@@ -1,0 +1,165 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"testing"
+
+	commonEvent "github.com/pingcap/ticdc/pkg/common/event"
+	"github.com/stretchr/testify/require"
+)
+
+// TestShouldGenBatchSQL tests the shouldGenBatchSQL function
+func TestShouldGenBatchSQL(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name           string
+		hasPK          bool
+		hasVirtualCols bool
+		events         []*commonEvent.DMLEvent
+		safemode       bool
+		want           bool
+	}{
+		{
+			name:           "table without primary key should not use batch SQL",
+			hasPK:          false,
+			hasVirtualCols: false,
+			events:         []*commonEvent.DMLEvent{newDMLEvent(t, 1, 1, 1)},
+			safemode:       false,
+			want:           false,
+		},
+		{
+			name:           "table with virtual columns should not use batch SQL",
+			hasPK:          true,
+			hasVirtualCols: true,
+			events:         []*commonEvent.DMLEvent{newDMLEvent(t, 1, 1, 1)},
+			safemode:       false,
+			want:           false,
+		},
+		{
+			name:           "single row event should not use batch SQL",
+			hasPK:          true,
+			hasVirtualCols: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 1, 1, 1),
+			},
+			safemode: false,
+			want:     false,
+		},
+		{
+			name:           "safe mode should not use batch SQL",
+			hasPK:          true,
+			hasVirtualCols: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 1, 2, 2),
+				newDMLEvent(t, 2, 3, 2),
+			},
+			safemode: false,
+			want:     false,
+		},
+		{
+			name:           "multiple rows with primary key in unsafe mode should use batch SQL",
+			hasPK:          true,
+			hasVirtualCols: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 2, 1, 2),
+				newDMLEvent(t, 3, 1, 2),
+			},
+			safemode: false,
+			want:     true,
+		},
+		{
+			name:           "global safe mode should not use batch SQL",
+			hasPK:          true,
+			hasVirtualCols: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 2, 1, 2),
+			},
+			safemode: true,
+			want:     false,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			got := shouldGenBatchSQL(tt.hasPK, tt.hasVirtualCols, tt.events, tt.safemode)
+			require.Equal(t, tt.want, got)
+		})
+	}
+}
+
+// TestShouldSafeMode tests the shouldSafeMode function
+func TestShouldSafeMode(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name     string
+		safemode bool
+		events   []*commonEvent.DMLEvent
+		want     bool
+	}{
+		{
+			name:     "global safe mode enabled",
+			safemode: true,
+			events:   []*commonEvent.DMLEvent{newDMLEvent(t, 2, 1, 1)},
+			want:     true,
+		},
+		{
+			name:     "all events have CommitTs > ReplicatingTs",
+			safemode: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 2, 1, 1),
+				newDMLEvent(t, 3, 2, 1),
+			},
+			want: false,
+		},
+		{
+			name:     "at least one event has CommitTs < ReplicatingTs",
+			safemode: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 2, 1, 1),
+				newDMLEvent(t, 1, 2, 1), // This event triggers safe mode
+				newDMLEvent(t, 3, 2, 1),
+			},
+			want: true,
+		},
+		{
+			name:     "empty events should not trigger safe mode",
+			safemode: false,
+			events:   []*commonEvent.DMLEvent{},
+			want:     false,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			got := shouldSafeMode(tt.safemode, tt.events)
+			require.Equal(t, tt.want, got)
+		})
+	}
+}
+
+// newDMLEvent creates a mock DMLEvent for testing
+func newDMLEvent(_ *testing.T, commitTs, replicatingTs, rowCount uint64) *commonEvent.DMLEvent {
+	return &commonEvent.DMLEvent{
+		CommitTs:      commitTs,
+		ReplicatingTs: replicatingTs,
+		Length:        int32(rowCount),
+	}
+}

--- a/pkg/sink/mysql/mysql_writer_dml_test.go
+++ b/pkg/sink/mysql/mysql_writer_dml_test.go
@@ -59,7 +59,7 @@ func TestShouldGenBatchSQL(t *testing.T) {
 			want:     false,
 		},
 		{
-			name:           "safe mode should not use batch SQL",
+			name:           "all rows in same safe mode should use batch SQL",
 			hasPK:          true,
 			hasVirtualCols: false,
 			events: []*commonEvent.DMLEvent{
@@ -67,7 +67,7 @@ func TestShouldGenBatchSQL(t *testing.T) {
 				newDMLEvent(t, 2, 3, 2),
 			},
 			safemode: false,
-			want:     false,
+			want:     true,
 		},
 		{
 			name:           "multiple rows with primary key in unsafe mode should use batch SQL",
@@ -81,13 +81,24 @@ func TestShouldGenBatchSQL(t *testing.T) {
 			want:     true,
 		},
 		{
-			name:           "global safe mode should not use batch SQL",
+			name:           "global safe mode should use batch SQL",
 			hasPK:          true,
 			hasVirtualCols: false,
 			events: []*commonEvent.DMLEvent{
 				newDMLEvent(t, 2, 1, 2),
 			},
 			safemode: true,
+			want:     true,
+		},
+		{
+			name:           "multiple rows with primary key in different safe mode should not use batch SQL",
+			hasPK:          true,
+			hasVirtualCols: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 2, 1, 2),
+				newDMLEvent(t, 1, 2, 2),
+			},
+			safemode: false,
 			want:     false,
 		},
 	}
@@ -102,8 +113,9 @@ func TestShouldGenBatchSQL(t *testing.T) {
 	}
 }
 
-// TestShouldSafeMode tests the shouldSafeMode function
-func TestShouldSafeMode(t *testing.T) {
+// TestAllRowInSameSafeMode tests the allRowInSameSafeMode function which determines
+// if all rows in a batch of DML events have the same safe mode status
+func TestAllRowInSameSafeMode(t *testing.T) {
 	t.Parallel()
 
 	tests := []struct {
@@ -119,29 +131,52 @@ func TestShouldSafeMode(t *testing.T) {
 			want:     true,
 		},
 		{
-			name:     "all events have CommitTs > ReplicatingTs",
+			name:     "all events have same safe mode status (all CommitTs > ReplicatingTs)",
 			safemode: false,
 			events: []*commonEvent.DMLEvent{
 				newDMLEvent(t, 2, 1, 1),
-				newDMLEvent(t, 3, 2, 1),
-			},
-			want: false,
-		},
-		{
-			name:     "at least one event has CommitTs < ReplicatingTs",
-			safemode: false,
-			events: []*commonEvent.DMLEvent{
-				newDMLEvent(t, 2, 1, 1),
-				newDMLEvent(t, 1, 2, 1), // This event triggers safe mode
 				newDMLEvent(t, 3, 2, 1),
 			},
 			want: true,
 		},
 		{
-			name:     "empty events should not trigger safe mode",
+			name:     "all events have same safe mode status (all CommitTs <= ReplicatingTs)",
+			safemode: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 1, 1, 1),
+				newDMLEvent(t, 1, 2, 1),
+			},
+			want: true,
+		},
+		{
+			name:     "events have mixed safe mode status",
+			safemode: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 2, 1, 1), // CommitTs > ReplicatingTs
+				newDMLEvent(t, 1, 2, 1), // CommitTs < ReplicatingTs
+			},
+			want: false,
+		},
+		{
+			name:     "events have mixed safe mode status (equal case)",
+			safemode: false,
+			events: []*commonEvent.DMLEvent{
+				newDMLEvent(t, 2, 1, 1), // CommitTs > ReplicatingTs
+				newDMLEvent(t, 2, 2, 1), // CommitTs = ReplicatingTs
+			},
+			want: false,
+		},
+		{
+			name:     "empty events array",
 			safemode: false,
 			events:   []*commonEvent.DMLEvent{},
 			want:     false,
+		},
+		{
+			name:     "single event",
+			safemode: false,
+			events:   []*commonEvent.DMLEvent{newDMLEvent(t, 1, 1, 1)},
+			want:     true,
 		},
 	}
 
@@ -149,7 +184,7 @@ func TestShouldSafeMode(t *testing.T) {
 		tt := tt
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
-			got := shouldSafeMode(tt.safemode, tt.events)
+			got := allRowInSameSafeMode(tt.safemode, tt.events)
 			require.Equal(t, tt.want, got)
 		})
 	}

